{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import sys"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In hg19, there are duplicate entries for a select few older files.  Newer entries may be distinguished by presence of a dateResubmitted field."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def readMetaDataIntoList(meta_file):\n",
    "    f = open(meta_file)\n",
    "    meta_dict = dict()\n",
    "    for line in f:\n",
    "        file_name = line.split(\"\\t\")[0]\n",
    "        if file_name not in meta_dict:\n",
    "            meta_dict[file_name] = []\n",
    "        entry_dict = dict()\n",
    "        for entry in line.split(\"\\t\")[1].split(\";\"):\n",
    "            entry_dict[entry.strip().split(\"=\")[0]] = entry.strip().split(\"=\")[1]\n",
    "        meta_dict[file_name].append(entry_dict)\n",
    "    return meta_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def checkUniqueFileNames(meta_dict):\n",
    "    for file_name in meta_dict:\n",
    "        if len(meta_dict[file_name]) > 1:\n",
    "            return False\n",
    "    return True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "hg19_meta_list = readMetaDataIntoList(\"hg19_metadata.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "False\n"
     ]
    }
   ],
   "source": [
    "print(checkUniqueFileNames(hg19_meta_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "wgEncodeBroadHistoneHuvecH3k27acStdSig.bigWig\n",
      "{'control': 'std', 'controlId': 'wgEncodeEH000060', 'project': 'wgEncode', 'type': 'bigWig', 'origAssembly': 'hg18', 'lab': 'Broad', 'dccAccession': 'wgEncodeEH000055', 'cell': 'HUVEC', 'tableName': 'wgEncodeBroadHistoneHuvecH3k27acStdSig', 'treatment': 'None', 'composite': 'wgEncodeBroadHistone', 'dataType': 'ChipSeq', 'geoSampleAccession': 'GSM733691', 'view': 'Signal', 'dateUnrestricted': '2009-10-06', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'md5sum': '5ab0cc0d6511dd9417a976c485b2ff57', 'antibody': 'H3K27ac', 'dateResubmitted': '2010-11-05', 'subId': '2874', 'size': '296M', 'grant': 'Bernstein', 'dateSubmitted': '2009-01-06'}\n",
      "{'cell': 'HUVEC', 'type': 'bigWig', 'dataType': 'ChipSeq', 'controlId': 'HUVEC/Input/std', 'dccAccession': 'wgEncodeEH000055', 'expId': '55', 'origAssembly': 'hg18', 'view': 'Signal', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'antibody': 'H3K27ac', 'dateResubmitted': '2010-11-05', 'lab': 'Broad', 'subId': '2874', 'size': '296M', 'grant': 'Bernstein', 'control': 'std', 'dateSubmitted': '2009-01-06'}\n",
      "wgEncodeBroadHistoneNhlfH3k4me3StdSig.bigWig\n",
      "{'control': 'std', 'controlId': 'wgEncodeEH000105', 'project': 'wgEncode', 'type': 'bigWig', 'origAssembly': 'hg18', 'lab': 'Broad', 'dccAccession': 'wgEncodeEH000102', 'cell': 'NHLF', 'tableName': 'wgEncodeBroadHistoneNhlfH3k4me3StdSig', 'treatment': 'None', 'composite': 'wgEncodeBroadHistone', 'dataType': 'ChipSeq', 'geoSampleAccession': 'GSM733723', 'view': 'Signal', 'dateUnrestricted': '2010-06-28', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'md5sum': 'dcc5f004bda539fcee6518a31a8e3916', 'antibody': 'H3K4me3', 'dateResubmitted': '2010-11-05', 'subId': '2921', 'size': '427M', 'grant': 'Bernstein', 'dateSubmitted': '2009-09-29'}\n",
      "{'cell': 'NHLF', 'type': 'bigWig', 'dataType': 'ChipSeq', 'controlId': 'NHLF/Input/std', 'dccAccession': 'wgEncodeEH000102', 'expId': '102', 'origAssembly': 'hg18', 'view': 'Signal', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'antibody': 'H3K4me3', 'dateResubmitted': '2010-11-05', 'lab': 'Broad', 'subId': '2921', 'size': '427M', 'grant': 'Bernstein', 'control': 'std', 'dateSubmitted': '2009-09-29'}\n",
      "wgEncodeBroadHistoneH1hescH3k27acStdSig.bigWig\n",
      "{'control': 'std', 'controlId': 'wgEncodeEH000088', 'project': 'wgEncode', 'type': 'bigWig', 'origAssembly': 'hg19', 'lab': 'Broad', 'dccAccession': 'wgEncodeEH000997', 'cell': 'H1-hESC', 'tableName': 'wgEncodeBroadHistoneH1hescH3k27acStdSig', 'treatment': 'None', 'composite': 'wgEncodeBroadHistone', 'dataType': 'ChipSeq', 'geoSampleAccession': 'GSM733718', 'view': 'Signal', 'dateUnrestricted': '2011-03-21', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'md5sum': 'e89e3efed7655c85adba862e2ef2fa11', 'antibody': 'H3K27ac', 'dateResubmitted': '2010-11-05', 'subId': '2811', 'size': '652M', 'grant': 'Bernstein', 'dateSubmitted': '2010-06-21'}\n",
      "{'cell': 'H1-hESC', 'type': 'bigWig', 'dataType': 'ChipSeq', 'controlId': 'H1-hESC/Input/std', 'dccAccession': 'wgEncodeEH000997', 'expId': '997', 'origAssembly': 'hg19', 'view': 'Signal', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'antibody': 'H3K27ac', 'dateResubmitted': '2010-11-05', 'lab': 'Broad', 'subId': '2811', 'size': '652M', 'grant': 'Bernstein', 'control': 'std', 'dateSubmitted': '2010-06-21'}\n",
      "wgEncodeBroadHistoneH1hescH3k4me3StdSig.bigWig\n",
      "{'control': 'std', 'controlId': 'wgEncodeEH000088', 'project': 'wgEncode', 'type': 'bigWig', 'origAssembly': 'hg18', 'lab': 'Broad', 'dccAccession': 'wgEncodeEH000086', 'cell': 'H1-hESC', 'tableName': 'wgEncodeBroadHistoneH1hescH3k4me3StdSig', 'treatment': 'None', 'composite': 'wgEncodeBroadHistone', 'dataType': 'ChipSeq', 'geoSampleAccession': 'GSM733657', 'view': 'Signal', 'dateUnrestricted': '2010-06-28', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'md5sum': '98d1150a5ce3e5aa17bb55c7a2fd2aa3', 'antibody': 'H3K4me3', 'dateResubmitted': '2010-11-05', 'subId': '2815', 'size': '320M', 'grant': 'Bernstein', 'dateSubmitted': '2009-09-29'}\n",
      "{'cell': 'H1-hESC', 'type': 'bigWig', 'dataType': 'ChipSeq', 'controlId': 'H1-hESC/Input/std', 'dccAccession': 'wgEncodeEH000086', 'expId': '86', 'origAssembly': 'hg18', 'view': 'Signal', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'antibody': 'H3K4me3', 'dateResubmitted': '2010-11-05', 'lab': 'Broad', 'subId': '2815', 'size': '320M', 'grant': 'Bernstein', 'control': 'std', 'dateSubmitted': '2009-09-29'}\n",
      "wgEncodeBroadHistoneK562H3k4me1StdSig.bigWig\n",
      "{'control': 'std', 'controlId': 'wgEncodeEH000052', 'project': 'wgEncode', 'type': 'bigWig', 'origAssembly': 'hg18', 'lab': 'Broad', 'dccAccession': 'wgEncodeEH000046', 'cell': 'K562', 'tableName': 'wgEncodeBroadHistoneK562H3k4me1StdSig', 'treatment': 'None', 'composite': 'wgEncodeBroadHistone', 'dataType': 'ChipSeq', 'geoSampleAccession': 'GSM733692', 'view': 'Signal', 'dateUnrestricted': '2009-10-05', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'md5sum': '00b5f5a38c21988b740ea4bfc626cde6', 'antibody': 'H3K4me1', 'dateResubmitted': '2010-11-05', 'subId': '2888', 'size': '355M', 'grant': 'Bernstein', 'dateSubmitted': '2009-01-05'}\n",
      "{'cell': 'K562', 'type': 'bigWig', 'dataType': 'ChipSeq', 'controlId': 'K562/Input/std', 'dccAccession': 'wgEncodeEH000046', 'expId': '46', 'origAssembly': 'hg18', 'view': 'Signal', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'antibody': 'H3K4me1', 'dateResubmitted': '2010-11-05', 'lab': 'Broad', 'subId': '2888', 'size': '355M', 'grant': 'Bernstein', 'control': 'std', 'dateSubmitted': '2009-01-05'}\n",
      "wgEncodeBroadHistoneHuvecH3k4me1StdSig.bigWig\n",
      "{'control': 'std', 'controlId': 'wgEncodeEH000060', 'project': 'wgEncode', 'type': 'bigWig', 'origAssembly': 'hg18', 'lab': 'Broad', 'dccAccession': 'wgEncodeEH000039', 'cell': 'HUVEC', 'tableName': 'wgEncodeBroadHistoneHuvecH3k4me1StdSig', 'treatment': 'None', 'composite': 'wgEncodeBroadHistone', 'dataType': 'ChipSeq', 'geoSampleAccession': 'GSM733690', 'view': 'Signal', 'dateUnrestricted': '2009-10-05', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'md5sum': '8c96b3374dd49be97c5087483807b151', 'antibody': 'H3K4me1', 'dateResubmitted': '2010-11-05', 'subId': '2876', 'size': '345M', 'grant': 'Bernstein', 'dateSubmitted': '2009-01-05'}\n",
      "{'cell': 'HUVEC', 'type': 'bigWig', 'dataType': 'ChipSeq', 'controlId': 'HUVEC/Input/std', 'dccAccession': 'wgEncodeEH000039', 'expId': '39', 'origAssembly': 'hg18', 'view': 'Signal', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'antibody': 'H3K4me1', 'dateResubmitted': '2010-11-05', 'lab': 'Broad', 'subId': '2876', 'size': '345M', 'grant': 'Bernstein', 'control': 'std', 'dateSubmitted': '2009-01-05'}\n",
      "wgEncodeBroadHistoneNhlfH3k4me1StdSig.bigWig\n",
      "{'control': 'std', 'controlId': 'wgEncodeEH000105', 'project': 'wgEncode', 'type': 'bigWig', 'origAssembly': 'hg18', 'lab': 'Broad', 'dccAccession': 'wgEncodeEH000100', 'cell': 'NHLF', 'tableName': 'wgEncodeBroadHistoneNhlfH3k4me1StdSig', 'treatment': 'None', 'composite': 'wgEncodeBroadHistone', 'dataType': 'ChipSeq', 'geoSampleAccession': 'GSM733649', 'view': 'Signal', 'dateUnrestricted': '2010-06-28', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'md5sum': '13e2a8d3849e9b2637f0bb9b17db27eb', 'antibody': 'H3K4me1', 'dateResubmitted': '2010-11-05', 'subId': '2919', 'size': '340M', 'grant': 'Bernstein', 'dateSubmitted': '2009-09-29'}\n",
      "{'cell': 'NHLF', 'type': 'bigWig', 'dataType': 'ChipSeq', 'controlId': 'NHLF/Input/std', 'dccAccession': 'wgEncodeEH000100', 'expId': '100', 'origAssembly': 'hg18', 'view': 'Signal', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'antibody': 'H3K4me1', 'dateResubmitted': '2010-11-05', 'lab': 'Broad', 'subId': '2919', 'size': '340M', 'grant': 'Bernstein', 'control': 'std', 'dateSubmitted': '2009-09-29'}\n",
      "wgEncodeBroadHistoneNhlfH3k27acStdSig.bigWig\n",
      "{'control': 'std', 'controlId': 'wgEncodeEH000105', 'project': 'wgEncode', 'type': 'bigWig', 'origAssembly': 'hg18', 'lab': 'Broad', 'dccAccession': 'wgEncodeEH000097', 'cell': 'NHLF', 'tableName': 'wgEncodeBroadHistoneNhlfH3k27acStdSig', 'treatment': 'None', 'composite': 'wgEncodeBroadHistone', 'dataType': 'ChipSeq', 'geoSampleAccession': 'GSM733646', 'view': 'Signal', 'dateUnrestricted': '2010-06-28', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'md5sum': 'b5711998f2002b84146f7e8a4f3db888', 'antibody': 'H3K27ac', 'dateResubmitted': '2010-11-05', 'subId': '2916', 'size': '259M', 'grant': 'Bernstein', 'dateSubmitted': '2009-09-29'}\n",
      "{'cell': 'NHLF', 'type': 'bigWig', 'dataType': 'ChipSeq', 'controlId': 'NHLF/Input/std', 'dccAccession': 'wgEncodeEH000097', 'expId': '97', 'origAssembly': 'hg18', 'view': 'Signal', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'antibody': 'H3K27ac', 'dateResubmitted': '2010-11-05', 'lab': 'Broad', 'subId': '2916', 'size': '259M', 'grant': 'Bernstein', 'control': 'std', 'dateSubmitted': '2009-09-29'}\n",
      "wgEncodeBroadHistoneHuvecH3k4me3StdSig.bigWig\n",
      "{'control': 'std', 'controlId': 'wgEncodeEH000060', 'project': 'wgEncode', 'type': 'bigWig', 'origAssembly': 'hg18', 'lab': 'Broad', 'dccAccession': 'wgEncodeEH000041', 'cell': 'HUVEC', 'tableName': 'wgEncodeBroadHistoneHuvecH3k4me3StdSig', 'treatment': 'None', 'composite': 'wgEncodeBroadHistone', 'dataType': 'ChipSeq', 'geoSampleAccession': 'GSM733673', 'view': 'Signal', 'dateUnrestricted': '2009-10-05', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'md5sum': 'f9d3af3fd6d4a3e3c70cd38a038e043a', 'antibody': 'H3K4me3', 'dateResubmitted': '2010-11-05', 'subId': '2878', 'size': '343M', 'grant': 'Bernstein', 'dateSubmitted': '2009-01-05'}\n",
      "{'cell': 'HUVEC', 'type': 'bigWig', 'dataType': 'ChipSeq', 'controlId': 'HUVEC/Input/std', 'dccAccession': 'wgEncodeEH000041', 'expId': '41', 'origAssembly': 'hg18', 'view': 'Signal', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'antibody': 'H3K4me3', 'dateResubmitted': '2010-11-05', 'lab': 'Broad', 'subId': '2878', 'size': '343M', 'grant': 'Bernstein', 'control': 'std', 'dateSubmitted': '2009-01-05'}\n",
      "wgEncodeBroadHistoneK562H3k27acStdSig.bigWig\n",
      "{'control': 'std', 'controlId': 'wgEncodeEH000052', 'project': 'wgEncode', 'type': 'bigWig', 'origAssembly': 'hg18', 'lab': 'Broad', 'dccAccession': 'wgEncodeEH000043', 'cell': 'K562', 'tableName': 'wgEncodeBroadHistoneK562H3k27acStdSig', 'treatment': 'None', 'composite': 'wgEncodeBroadHistone', 'dataType': 'ChipSeq', 'geoSampleAccession': 'GSM733656', 'view': 'Signal', 'dateUnrestricted': '2009-10-05', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'md5sum': '375b66adfebf3c4eab113fb4a449adf2', 'antibody': 'H3K27ac', 'dateResubmitted': '2010-11-05', 'subId': '2885', 'size': '293M', 'grant': 'Bernstein', 'dateSubmitted': '2009-01-05'}\n",
      "{'cell': 'K562', 'type': 'bigWig', 'dataType': 'ChipSeq', 'controlId': 'K562/Input/std', 'dccAccession': 'wgEncodeEH000043', 'expId': '43', 'origAssembly': 'hg18', 'view': 'Signal', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'antibody': 'H3K27ac', 'dateResubmitted': '2010-11-05', 'lab': 'Broad', 'subId': '2885', 'size': '293M', 'grant': 'Bernstein', 'control': 'std', 'dateSubmitted': '2009-01-05'}\n",
      "wgEncodeBroadHistoneNhekH3k4me1StdSig.bigWig\n",
      "{'control': 'std', 'controlId': 'wgEncodeEH000072', 'project': 'wgEncode', 'type': 'bigWig', 'origAssembly': 'hg18', 'lab': 'Broad', 'dccAccession': 'wgEncodeEH000062', 'cell': 'NHEK', 'tableName': 'wgEncodeBroadHistoneNhekH3k4me1StdSig', 'treatment': 'None', 'composite': 'wgEncodeBroadHistone', 'dataType': 'ChipSeq', 'geoSampleAccession': 'GSM733698', 'view': 'Signal', 'dateUnrestricted': '2009-10-06', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'md5sum': '3613000e4bdc0c01332266fcce008eb9', 'antibody': 'H3K4me1', 'dateResubmitted': '2010-11-05', 'subId': '2907', 'size': '301M', 'grant': 'Bernstein', 'dateSubmitted': '2009-01-06'}\n",
      "{'cell': 'NHEK', 'type': 'bigWig', 'dataType': 'ChipSeq', 'controlId': 'NHEK/Input/std', 'dccAccession': 'wgEncodeEH000062', 'expId': '62', 'origAssembly': 'hg18', 'view': 'Signal', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'antibody': 'H3K4me1', 'dateResubmitted': '2010-11-05', 'lab': 'Broad', 'subId': '2907', 'size': '301M', 'grant': 'Bernstein', 'control': 'std', 'dateSubmitted': '2009-01-06'}\n",
      "wgEncodeBroadHistoneH1hescH3k4me1StdSig.bigWig\n",
      "{'control': 'std', 'controlId': 'wgEncodeEH000088', 'project': 'wgEncode', 'type': 'bigWig', 'origAssembly': 'hg18', 'lab': 'Broad', 'dccAccession': 'wgEncodeEH000106', 'cell': 'H1-hESC', 'tableName': 'wgEncodeBroadHistoneH1hescH3k4me1StdSig', 'treatment': 'None', 'composite': 'wgEncodeBroadHistone', 'dataType': 'ChipSeq', 'geoSampleAccession': 'GSM733782', 'view': 'Signal', 'dateUnrestricted': '2010-06-30', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'md5sum': '82655f8e6a8cec482bb03f3a1810328d', 'antibody': 'H3K4me1', 'dateResubmitted': '2010-11-05', 'subId': '2796', 'size': '411M', 'grant': 'Bernstein', 'dateSubmitted': '2009-09-30'}\n",
      "{'cell': 'H1-hESC', 'type': 'bigWig', 'dataType': 'ChipSeq', 'controlId': 'H1-hESC/Input/std', 'dccAccession': 'wgEncodeEH000106', 'expId': '106', 'origAssembly': 'hg18', 'view': 'Signal', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'antibody': 'H3K4me1', 'dateResubmitted': '2010-11-05', 'lab': 'Broad', 'subId': '2796', 'size': '411M', 'grant': 'Bernstein', 'control': 'std', 'dateSubmitted': '2009-09-30'}\n",
      "wgEncodeBroadHistoneHsmmH3k4me1StdSig.bigWig\n",
      "{'control': 'std', 'controlId': 'wgEncodeEH000119', 'project': 'wgEncode', 'type': 'bigWig', 'origAssembly': 'hg18', 'lab': 'Broad', 'dccAccession': 'wgEncodeEH000114', 'cell': 'HSMM', 'tableName': 'wgEncodeBroadHistoneHsmmH3k4me1StdSig', 'treatment': 'None', 'composite': 'wgEncodeBroadHistone', 'dataType': 'ChipSeq', 'geoSampleAccession': 'GSM733761', 'view': 'Signal', 'dateUnrestricted': '2010-09-16', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'md5sum': 'ced361460f1038381aca10de7e142bd7', 'antibody': 'H3K4me1', 'dateResubmitted': '2010-11-05', 'subId': '2844', 'size': '330M', 'grant': 'Bernstein', 'dateSubmitted': '2009-12-16'}\n",
      "{'cell': 'HSMM', 'type': 'bigWig', 'dataType': 'ChipSeq', 'controlId': 'HSMM/Input/std', 'dccAccession': 'wgEncodeEH000114', 'expId': '114', 'origAssembly': 'hg18', 'view': 'Signal', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'antibody': 'H3K4me1', 'dateResubmitted': '2010-11-05', 'lab': 'Broad', 'subId': '2844', 'size': '330M', 'grant': 'Bernstein', 'control': 'std', 'dateSubmitted': '2009-12-16'}\n",
      "wgEncodeBroadHistoneGm12878H3k4me1StdSig.bigWig\n",
      "{'control': 'std', 'controlId': 'wgEncodeEH000037', 'project': 'wgEncode', 'type': 'bigWig', 'objStatus': 'replaced - Reanalyzed after additional sequencing of first replicate added', 'origAssembly': 'hg18', 'lab': 'Broad', 'dccAccession': 'wgEncodeEH000033', 'cell': 'GM12878', 'tableName': 'wgEncodeBroadHistoneGm12878H3k4me1StdSig', 'treatment': 'None', 'composite': 'wgEncodeBroadHistone', 'dataType': 'ChipSeq', 'md5sum': '31e171a083ee87d12d6f7f9a6fe34897', 'view': 'Signal', 'dateUnrestricted': '2009-10-05', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'antibody': 'H3K4me1', 'dateResubmitted': '2010-11-05', 'subId': '2804', 'size': '346M', 'grant': 'Bernstein', 'dateSubmitted': '2009-01-05'}\n",
      "{'cell': 'GM12878', 'type': 'bigWig', 'dataType': 'ChipSeq', 'controlId': 'GM12878/Input/std', 'dccAccession': 'wgEncodeEH000033', 'expId': '33', 'origAssembly': 'hg18', 'view': 'Signal', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'antibody': 'H3K4me1', 'dateResubmitted': '2010-11-05', 'lab': 'Broad', 'subId': '2804', 'size': '346M', 'grant': 'Bernstein', 'control': 'std', 'dateSubmitted': '2009-01-05'}\n",
      "wgEncodeBroadHistoneHsmmH3k4me3StdSig.bigWig\n",
      "{'control': 'std', 'controlId': 'wgEncodeEH000119', 'project': 'wgEncode', 'type': 'bigWig', 'origAssembly': 'hg18', 'lab': 'Broad', 'dccAccession': 'wgEncodeEH000116', 'cell': 'HSMM', 'tableName': 'wgEncodeBroadHistoneHsmmH3k4me3StdSig', 'treatment': 'None', 'composite': 'wgEncodeBroadHistone', 'dataType': 'ChipSeq', 'geoSampleAccession': 'GSM733637', 'view': 'Signal', 'dateUnrestricted': '2010-09-16', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'md5sum': 'b9a1ecd6a2de34b7fdf1fc78d1ffddbb', 'antibody': 'H3K4me3', 'dateResubmitted': '2010-11-05', 'subId': '2846', 'size': '258M', 'grant': 'Bernstein', 'dateSubmitted': '2009-12-16'}\n",
      "{'cell': 'HSMM', 'type': 'bigWig', 'dataType': 'ChipSeq', 'controlId': 'HSMM/Input/std', 'dccAccession': 'wgEncodeEH000116', 'expId': '116', 'origAssembly': 'hg18', 'view': 'Signal', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'antibody': 'H3K4me3', 'dateResubmitted': '2010-11-05', 'lab': 'Broad', 'subId': '2846', 'size': '258M', 'grant': 'Bernstein', 'control': 'std', 'dateSubmitted': '2009-12-16'}\n",
      "wgEncodeBroadHistoneGm12878H3k27acStdSig.bigWig\n",
      "{'control': 'std', 'controlId': 'wgEncodeEH000037', 'project': 'wgEncode', 'type': 'bigWig', 'origAssembly': 'hg18', 'lab': 'Broad', 'dccAccession': 'wgEncodeEH000030', 'cell': 'GM12878', 'tableName': 'wgEncodeBroadHistoneGm12878H3k27acStdSig', 'treatment': 'None', 'composite': 'wgEncodeBroadHistone', 'dataType': 'ChipSeq', 'geoSampleAccession': 'GSM733771', 'view': 'Signal', 'dateUnrestricted': '2009-10-05', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'md5sum': '4fc2dec4b72bc85f6087be52267beec2', 'antibody': 'H3K27ac', 'dateResubmitted': '2010-11-05', 'subId': '2801', 'size': '250M', 'grant': 'Bernstein', 'dateSubmitted': '2009-01-05'}\n",
      "{'cell': 'GM12878', 'type': 'bigWig', 'dataType': 'ChipSeq', 'controlId': 'GM12878/Input/std', 'dccAccession': 'wgEncodeEH000030', 'expId': '30', 'origAssembly': 'hg18', 'view': 'Signal', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'antibody': 'H3K27ac', 'dateResubmitted': '2010-11-05', 'lab': 'Broad', 'subId': '2801', 'size': '250M', 'grant': 'Bernstein', 'control': 'std', 'dateSubmitted': '2009-01-05'}\n",
      "wgEncodeBroadHistoneNhekH3k4me3StdSig.bigWig\n",
      "{'control': 'std', 'controlId': 'wgEncodeEH000072', 'project': 'wgEncode', 'type': 'bigWig', 'origAssembly': 'hg18', 'lab': 'Broad', 'dccAccession': 'wgEncodeEH000068', 'cell': 'NHEK', 'tableName': 'wgEncodeBroadHistoneNhekH3k4me3StdSig', 'treatment': 'None', 'composite': 'wgEncodeBroadHistone', 'dataType': 'ChipSeq', 'geoSampleAccession': 'GSM733720', 'view': 'Signal', 'dateUnrestricted': '2009-10-07', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'md5sum': 'ca1ffa08c5152fc70bd5523dbbd71eb7', 'antibody': 'H3K4me3', 'dateResubmitted': '2010-11-05', 'subId': '2909', 'size': '332M', 'grant': 'Bernstein', 'dateSubmitted': '2009-01-07'}\n",
      "{'cell': 'NHEK', 'type': 'bigWig', 'dataType': 'ChipSeq', 'controlId': 'NHEK/Input/std', 'dccAccession': 'wgEncodeEH000068', 'expId': '68', 'origAssembly': 'hg18', 'view': 'Signal', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'antibody': 'H3K4me3', 'dateResubmitted': '2010-11-05', 'lab': 'Broad', 'subId': '2909', 'size': '332M', 'grant': 'Bernstein', 'control': 'std', 'dateSubmitted': '2009-01-07'}\n",
      "wgEncodeBroadHistoneK562H3k4me3StdSig.bigWig\n",
      "{'control': 'std', 'controlId': 'wgEncodeEH000052', 'project': 'wgEncode', 'type': 'bigWig', 'origAssembly': 'hg18', 'lab': 'Broad', 'dccAccession': 'wgEncodeEH000048', 'cell': 'K562', 'tableName': 'wgEncodeBroadHistoneK562H3k4me3StdSig', 'treatment': 'None', 'composite': 'wgEncodeBroadHistone', 'dataType': 'ChipSeq', 'geoSampleAccession': 'GSM733680', 'view': 'Signal', 'dateUnrestricted': '2009-10-05', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'md5sum': '96646e11fb2f33384ca6d625eb8cc03b', 'antibody': 'H3K4me3', 'dateResubmitted': '2010-11-05', 'subId': '2890', 'size': '354M', 'grant': 'Bernstein', 'dateSubmitted': '2009-01-05'}\n",
      "{'cell': 'K562', 'type': 'bigWig', 'dataType': 'ChipSeq', 'controlId': 'K562/Input/std', 'dccAccession': 'wgEncodeEH000048', 'expId': '48', 'origAssembly': 'hg18', 'view': 'Signal', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'antibody': 'H3K4me3', 'dateResubmitted': '2010-11-05', 'lab': 'Broad', 'subId': '2890', 'size': '354M', 'grant': 'Bernstein', 'control': 'std', 'dateSubmitted': '2009-01-05'}\n",
      "wgEncodeBroadHistoneGm12878H3k4me3StdSig.bigWig\n",
      "{'control': 'std', 'controlId': 'wgEncodeEH000037', 'project': 'wgEncode', 'type': 'bigWig', 'objStatus': 'replaced - Reanalyzed after additional sequencing of second replicate added', 'origAssembly': 'hg18', 'lab': 'Broad', 'dccAccession': 'wgEncodeEH000028', 'cell': 'GM12878', 'tableName': 'wgEncodeBroadHistoneGm12878H3k4me3StdSig', 'treatment': 'None', 'composite': 'wgEncodeBroadHistone', 'dataType': 'ChipSeq', 'md5sum': 'a3b3600f038dfaf4ed6eb1724ffe19bf', 'view': 'Signal', 'dateUnrestricted': '2009-10-04', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'antibody': 'H3K4me3', 'dateResubmitted': '2010-11-05', 'subId': '2806', 'size': '435M', 'grant': 'Bernstein', 'dateSubmitted': '2009-01-04'}\n",
      "{'cell': 'GM12878', 'type': 'bigWig', 'dataType': 'ChipSeq', 'controlId': 'GM12878/Input/std', 'dccAccession': 'wgEncodeEH000028', 'expId': '28', 'origAssembly': 'hg18', 'view': 'Signal', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'antibody': 'H3K4me3', 'dateResubmitted': '2010-11-05', 'lab': 'Broad', 'subId': '2806', 'size': '435M', 'grant': 'Bernstein', 'control': 'std', 'dateSubmitted': '2009-01-04'}\n",
      "wgEncodeBroadHistoneNhekH3k27acStdSig.bigWig\n",
      "{'control': 'std', 'controlId': 'wgEncodeEH000072', 'project': 'wgEncode', 'type': 'bigWig', 'origAssembly': 'hg18', 'lab': 'Broad', 'dccAccession': 'wgEncodeEH000064', 'cell': 'NHEK', 'tableName': 'wgEncodeBroadHistoneNhekH3k27acStdSig', 'treatment': 'None', 'composite': 'wgEncodeBroadHistone', 'dataType': 'ChipSeq', 'geoSampleAccession': 'GSM733674', 'view': 'Signal', 'dateUnrestricted': '2009-10-07', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'md5sum': 'b9a22d33ccb14faa8cc97d042c78cc42', 'antibody': 'H3K27ac', 'dateResubmitted': '2010-11-05', 'subId': '2904', 'size': '280M', 'grant': 'Bernstein', 'dateSubmitted': '2009-01-07'}\n",
      "{'cell': 'NHEK', 'type': 'bigWig', 'dataType': 'ChipSeq', 'controlId': 'NHEK/Input/std', 'dccAccession': 'wgEncodeEH000064', 'expId': '64', 'origAssembly': 'hg18', 'view': 'Signal', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'antibody': 'H3K27ac', 'dateResubmitted': '2010-11-05', 'lab': 'Broad', 'subId': '2904', 'size': '280M', 'grant': 'Bernstein', 'control': 'std', 'dateSubmitted': '2009-01-07'}\n",
      "wgEncodeBroadHistoneHsmmH3k27acStdSig.bigWig\n",
      "{'control': 'std', 'controlId': 'wgEncodeEH000119', 'project': 'wgEncode', 'type': 'bigWig', 'origAssembly': 'hg18', 'lab': 'Broad', 'dccAccession': 'wgEncodeEH000111', 'cell': 'HSMM', 'tableName': 'wgEncodeBroadHistoneHsmmH3k27acStdSig', 'treatment': 'None', 'composite': 'wgEncodeBroadHistone', 'dataType': 'ChipSeq', 'geoSampleAccession': 'GSM733755', 'view': 'Signal', 'dateUnrestricted': '2010-09-16', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'md5sum': 'b2e50f27878d1927e6d0295d4ae2c0d3', 'antibody': 'H3K27ac', 'dateResubmitted': '2010-11-05', 'subId': '2841', 'size': '307M', 'grant': 'Bernstein', 'dateSubmitted': '2009-12-16'}\n",
      "{'cell': 'HSMM', 'type': 'bigWig', 'dataType': 'ChipSeq', 'controlId': 'HSMM/Input/std', 'dccAccession': 'wgEncodeEH000111', 'expId': '111', 'origAssembly': 'hg18', 'view': 'Signal', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'setType': 'exp', 'antibody': 'H3K27ac', 'dateResubmitted': '2010-11-05', 'lab': 'Broad', 'subId': '2841', 'size': '307M', 'grant': 'Bernstein', 'control': 'std', 'dateSubmitted': '2009-12-16'}\n"
     ]
    }
   ],
   "source": [
    "for file_name in hg19_meta_list:\n",
    "    if len(hg19_meta_list[file_name]) > 1:\n",
    "        print(file_name)\n",
    "        for entry_dict in hg19_meta_list[file_name]:\n",
    "            print(entry_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "wgEncodeBroadHistoneHuvecH3k27acStdSig.bigWig 0\n",
      "control: Same\n",
      "controlId: Different\n",
      "project: Not shared.\n",
      "type: Same\n",
      "origAssembly: Same\n",
      "lab: Same\n",
      "dccAccession: Same\n",
      "cell: Same\n",
      "tableName: Not shared.\n",
      "treatment: Not shared.\n",
      "composite: Not shared.\n",
      "dataType: Same\n",
      "geoSampleAccession: Not shared.\n",
      "view: Same\n",
      "dateUnrestricted: Not shared.\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "md5sum: Not shared.\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "dateSubmitted: Same\n",
      "wgEncodeBroadHistoneHuvecH3k27acStdSig.bigWig 1\n",
      "cell: Same\n",
      "type: Same\n",
      "dataType: Same\n",
      "controlId: Different\n",
      "dccAccession: Same\n",
      "expId: Not shared.\n",
      "origAssembly: Same\n",
      "view: Same\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "lab: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "control: Same\n",
      "dateSubmitted: Same\n",
      "wgEncodeBroadHistoneNhlfH3k4me3StdSig.bigWig 0\n",
      "control: Same\n",
      "controlId: Different\n",
      "project: Not shared.\n",
      "type: Same\n",
      "origAssembly: Same\n",
      "lab: Same\n",
      "dccAccession: Same\n",
      "cell: Same\n",
      "tableName: Not shared.\n",
      "treatment: Not shared.\n",
      "composite: Not shared.\n",
      "dataType: Same\n",
      "geoSampleAccession: Not shared.\n",
      "view: Same\n",
      "dateUnrestricted: Not shared.\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "md5sum: Not shared.\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "dateSubmitted: Same\n",
      "wgEncodeBroadHistoneNhlfH3k4me3StdSig.bigWig 1\n",
      "cell: Same\n",
      "type: Same\n",
      "dataType: Same\n",
      "controlId: Different\n",
      "dccAccession: Same\n",
      "expId: Not shared.\n",
      "origAssembly: Same\n",
      "view: Same\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "lab: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "control: Same\n",
      "dateSubmitted: Same\n",
      "wgEncodeBroadHistoneH1hescH3k27acStdSig.bigWig 0\n",
      "control: Same\n",
      "controlId: Different\n",
      "project: Not shared.\n",
      "type: Same\n",
      "origAssembly: Same\n",
      "lab: Same\n",
      "dccAccession: Same\n",
      "cell: Same\n",
      "tableName: Not shared.\n",
      "treatment: Not shared.\n",
      "composite: Not shared.\n",
      "dataType: Same\n",
      "geoSampleAccession: Not shared.\n",
      "view: Same\n",
      "dateUnrestricted: Not shared.\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "md5sum: Not shared.\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "dateSubmitted: Same\n",
      "wgEncodeBroadHistoneH1hescH3k27acStdSig.bigWig 1\n",
      "cell: Same\n",
      "type: Same\n",
      "dataType: Same\n",
      "controlId: Different\n",
      "dccAccession: Same\n",
      "expId: Not shared.\n",
      "origAssembly: Same\n",
      "view: Same\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "lab: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "control: Same\n",
      "dateSubmitted: Same\n",
      "wgEncodeBroadHistoneH1hescH3k4me3StdSig.bigWig 0\n",
      "control: Same\n",
      "controlId: Different\n",
      "project: Not shared.\n",
      "type: Same\n",
      "origAssembly: Same\n",
      "lab: Same\n",
      "dccAccession: Same\n",
      "cell: Same\n",
      "tableName: Not shared.\n",
      "treatment: Not shared.\n",
      "composite: Not shared.\n",
      "dataType: Same\n",
      "geoSampleAccession: Not shared.\n",
      "view: Same\n",
      "dateUnrestricted: Not shared.\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "md5sum: Not shared.\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "dateSubmitted: Same\n",
      "wgEncodeBroadHistoneH1hescH3k4me3StdSig.bigWig 1\n",
      "cell: Same\n",
      "type: Same\n",
      "dataType: Same\n",
      "controlId: Different\n",
      "dccAccession: Same\n",
      "expId: Not shared.\n",
      "origAssembly: Same\n",
      "view: Same\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "lab: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "control: Same\n",
      "dateSubmitted: Same\n",
      "wgEncodeBroadHistoneK562H3k4me1StdSig.bigWig 0\n",
      "control: Same\n",
      "controlId: Different\n",
      "project: Not shared.\n",
      "type: Same\n",
      "origAssembly: Same\n",
      "lab: Same\n",
      "dccAccession: Same\n",
      "cell: Same\n",
      "tableName: Not shared.\n",
      "treatment: Not shared.\n",
      "composite: Not shared.\n",
      "dataType: Same\n",
      "geoSampleAccession: Not shared.\n",
      "view: Same\n",
      "dateUnrestricted: Not shared.\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "md5sum: Not shared.\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "dateSubmitted: Same\n",
      "wgEncodeBroadHistoneK562H3k4me1StdSig.bigWig 1\n",
      "cell: Same\n",
      "type: Same\n",
      "dataType: Same\n",
      "controlId: Different\n",
      "dccAccession: Same\n",
      "expId: Not shared.\n",
      "origAssembly: Same\n",
      "view: Same\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "lab: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "control: Same\n",
      "dateSubmitted: Same\n",
      "wgEncodeBroadHistoneHuvecH3k4me1StdSig.bigWig 0\n",
      "control: Same\n",
      "controlId: Different\n",
      "project: Not shared.\n",
      "type: Same\n",
      "origAssembly: Same\n",
      "lab: Same\n",
      "dccAccession: Same\n",
      "cell: Same\n",
      "tableName: Not shared.\n",
      "treatment: Not shared.\n",
      "composite: Not shared.\n",
      "dataType: Same\n",
      "geoSampleAccession: Not shared.\n",
      "view: Same\n",
      "dateUnrestricted: Not shared.\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "md5sum: Not shared.\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "dateSubmitted: Same\n",
      "wgEncodeBroadHistoneHuvecH3k4me1StdSig.bigWig 1\n",
      "cell: Same\n",
      "type: Same\n",
      "dataType: Same\n",
      "controlId: Different\n",
      "dccAccession: Same\n",
      "expId: Not shared.\n",
      "origAssembly: Same\n",
      "view: Same\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "lab: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "control: Same\n",
      "dateSubmitted: Same\n",
      "wgEncodeBroadHistoneNhlfH3k4me1StdSig.bigWig 0\n",
      "control: Same\n",
      "controlId: Different\n",
      "project: Not shared.\n",
      "type: Same\n",
      "origAssembly: Same\n",
      "lab: Same\n",
      "dccAccession: Same\n",
      "cell: Same\n",
      "tableName: Not shared.\n",
      "treatment: Not shared.\n",
      "composite: Not shared.\n",
      "dataType: Same\n",
      "geoSampleAccession: Not shared.\n",
      "view: Same\n",
      "dateUnrestricted: Not shared.\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "md5sum: Not shared.\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "dateSubmitted: Same\n",
      "wgEncodeBroadHistoneNhlfH3k4me1StdSig.bigWig 1\n",
      "cell: Same\n",
      "type: Same\n",
      "dataType: Same\n",
      "controlId: Different\n",
      "dccAccession: Same\n",
      "expId: Not shared.\n",
      "origAssembly: Same\n",
      "view: Same\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "lab: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "control: Same\n",
      "dateSubmitted: Same\n",
      "wgEncodeBroadHistoneNhlfH3k27acStdSig.bigWig 0\n",
      "control: Same\n",
      "controlId: Different\n",
      "project: Not shared.\n",
      "type: Same\n",
      "origAssembly: Same\n",
      "lab: Same\n",
      "dccAccession: Same\n",
      "cell: Same\n",
      "tableName: Not shared.\n",
      "treatment: Not shared.\n",
      "composite: Not shared.\n",
      "dataType: Same\n",
      "geoSampleAccession: Not shared.\n",
      "view: Same\n",
      "dateUnrestricted: Not shared.\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "md5sum: Not shared.\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "dateSubmitted: Same\n",
      "wgEncodeBroadHistoneNhlfH3k27acStdSig.bigWig 1\n",
      "cell: Same\n",
      "type: Same\n",
      "dataType: Same\n",
      "controlId: Different\n",
      "dccAccession: Same\n",
      "expId: Not shared.\n",
      "origAssembly: Same\n",
      "view: Same\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "lab: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "control: Same\n",
      "dateSubmitted: Same\n",
      "wgEncodeBroadHistoneHuvecH3k4me3StdSig.bigWig 0\n",
      "control: Same\n",
      "controlId: Different\n",
      "project: Not shared.\n",
      "type: Same\n",
      "origAssembly: Same\n",
      "lab: Same\n",
      "dccAccession: Same\n",
      "cell: Same\n",
      "tableName: Not shared.\n",
      "treatment: Not shared.\n",
      "composite: Not shared.\n",
      "dataType: Same\n",
      "geoSampleAccession: Not shared.\n",
      "view: Same\n",
      "dateUnrestricted: Not shared.\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "md5sum: Not shared.\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "dateSubmitted: Same\n",
      "wgEncodeBroadHistoneHuvecH3k4me3StdSig.bigWig 1\n",
      "cell: Same\n",
      "type: Same\n",
      "dataType: Same\n",
      "controlId: Different\n",
      "dccAccession: Same\n",
      "expId: Not shared.\n",
      "origAssembly: Same\n",
      "view: Same\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "lab: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "control: Same\n",
      "dateSubmitted: Same\n",
      "wgEncodeBroadHistoneK562H3k27acStdSig.bigWig 0\n",
      "control: Same\n",
      "controlId: Different\n",
      "project: Not shared.\n",
      "type: Same\n",
      "origAssembly: Same\n",
      "lab: Same\n",
      "dccAccession: Same\n",
      "cell: Same\n",
      "tableName: Not shared.\n",
      "treatment: Not shared.\n",
      "composite: Not shared.\n",
      "dataType: Same\n",
      "geoSampleAccession: Not shared.\n",
      "view: Same\n",
      "dateUnrestricted: Not shared.\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "md5sum: Not shared.\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "dateSubmitted: Same\n",
      "wgEncodeBroadHistoneK562H3k27acStdSig.bigWig 1\n",
      "cell: Same\n",
      "type: Same\n",
      "dataType: Same\n",
      "controlId: Different\n",
      "dccAccession: Same\n",
      "expId: Not shared.\n",
      "origAssembly: Same\n",
      "view: Same\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "lab: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "control: Same\n",
      "dateSubmitted: Same\n",
      "wgEncodeBroadHistoneNhekH3k4me1StdSig.bigWig 0\n",
      "control: Same\n",
      "controlId: Different\n",
      "project: Not shared.\n",
      "type: Same\n",
      "origAssembly: Same\n",
      "lab: Same\n",
      "dccAccession: Same\n",
      "cell: Same\n",
      "tableName: Not shared.\n",
      "treatment: Not shared.\n",
      "composite: Not shared.\n",
      "dataType: Same\n",
      "geoSampleAccession: Not shared.\n",
      "view: Same\n",
      "dateUnrestricted: Not shared.\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "md5sum: Not shared.\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "dateSubmitted: Same\n",
      "wgEncodeBroadHistoneNhekH3k4me1StdSig.bigWig 1\n",
      "cell: Same\n",
      "type: Same\n",
      "dataType: Same\n",
      "controlId: Different\n",
      "dccAccession: Same\n",
      "expId: Not shared.\n",
      "origAssembly: Same\n",
      "view: Same\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "lab: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "control: Same\n",
      "dateSubmitted: Same\n",
      "wgEncodeBroadHistoneH1hescH3k4me1StdSig.bigWig 0\n",
      "control: Same\n",
      "controlId: Different\n",
      "project: Not shared.\n",
      "type: Same\n",
      "origAssembly: Same\n",
      "lab: Same\n",
      "dccAccession: Same\n",
      "cell: Same\n",
      "tableName: Not shared.\n",
      "treatment: Not shared.\n",
      "composite: Not shared.\n",
      "dataType: Same\n",
      "geoSampleAccession: Not shared.\n",
      "view: Same\n",
      "dateUnrestricted: Not shared.\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "md5sum: Not shared.\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "dateSubmitted: Same\n",
      "wgEncodeBroadHistoneH1hescH3k4me1StdSig.bigWig 1\n",
      "cell: Same\n",
      "type: Same\n",
      "dataType: Same\n",
      "controlId: Different\n",
      "dccAccession: Same\n",
      "expId: Not shared.\n",
      "origAssembly: Same\n",
      "view: Same\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "lab: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "control: Same\n",
      "dateSubmitted: Same\n",
      "wgEncodeBroadHistoneHsmmH3k4me1StdSig.bigWig 0\n",
      "control: Same\n",
      "controlId: Different\n",
      "project: Not shared.\n",
      "type: Same\n",
      "origAssembly: Same\n",
      "lab: Same\n",
      "dccAccession: Same\n",
      "cell: Same\n",
      "tableName: Not shared.\n",
      "treatment: Not shared.\n",
      "composite: Not shared.\n",
      "dataType: Same\n",
      "geoSampleAccession: Not shared.\n",
      "view: Same\n",
      "dateUnrestricted: Not shared.\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "md5sum: Not shared.\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "dateSubmitted: Same\n",
      "wgEncodeBroadHistoneHsmmH3k4me1StdSig.bigWig 1\n",
      "cell: Same\n",
      "type: Same\n",
      "dataType: Same\n",
      "controlId: Different\n",
      "dccAccession: Same\n",
      "expId: Not shared.\n",
      "origAssembly: Same\n",
      "view: Same\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "lab: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "control: Same\n",
      "dateSubmitted: Same\n",
      "wgEncodeBroadHistoneGm12878H3k4me1StdSig.bigWig 0\n",
      "control: Same\n",
      "controlId: Different\n",
      "project: Not shared.\n",
      "type: Same\n",
      "objStatus: Not shared.\n",
      "origAssembly: Same\n",
      "lab: Same\n",
      "dccAccession: Same\n",
      "cell: Same\n",
      "tableName: Not shared.\n",
      "treatment: Not shared.\n",
      "composite: Not shared.\n",
      "dataType: Same\n",
      "md5sum: Not shared.\n",
      "view: Same\n",
      "dateUnrestricted: Not shared.\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "dateSubmitted: Same\n",
      "wgEncodeBroadHistoneGm12878H3k4me1StdSig.bigWig 1\n",
      "cell: Same\n",
      "type: Same\n",
      "dataType: Same\n",
      "controlId: Different\n",
      "dccAccession: Same\n",
      "expId: Not shared.\n",
      "origAssembly: Same\n",
      "view: Same\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "lab: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "control: Same\n",
      "dateSubmitted: Same\n",
      "wgEncodeBroadHistoneHsmmH3k4me3StdSig.bigWig 0\n",
      "control: Same\n",
      "controlId: Different\n",
      "project: Not shared.\n",
      "type: Same\n",
      "origAssembly: Same\n",
      "lab: Same\n",
      "dccAccession: Same\n",
      "cell: Same\n",
      "tableName: Not shared.\n",
      "treatment: Not shared.\n",
      "composite: Not shared.\n",
      "dataType: Same\n",
      "geoSampleAccession: Not shared.\n",
      "view: Same\n",
      "dateUnrestricted: Not shared.\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "md5sum: Not shared.\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "dateSubmitted: Same\n",
      "wgEncodeBroadHistoneHsmmH3k4me3StdSig.bigWig 1\n",
      "cell: Same\n",
      "type: Same\n",
      "dataType: Same\n",
      "controlId: Different\n",
      "dccAccession: Same\n",
      "expId: Not shared.\n",
      "origAssembly: Same\n",
      "view: Same\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "lab: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "control: Same\n",
      "dateSubmitted: Same\n",
      "wgEncodeBroadHistoneGm12878H3k27acStdSig.bigWig 0\n",
      "control: Same\n",
      "controlId: Different\n",
      "project: Not shared.\n",
      "type: Same\n",
      "origAssembly: Same\n",
      "lab: Same\n",
      "dccAccession: Same\n",
      "cell: Same\n",
      "tableName: Not shared.\n",
      "treatment: Not shared.\n",
      "composite: Not shared.\n",
      "dataType: Same\n",
      "geoSampleAccession: Not shared.\n",
      "view: Same\n",
      "dateUnrestricted: Not shared.\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "md5sum: Not shared.\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "dateSubmitted: Same\n",
      "wgEncodeBroadHistoneGm12878H3k27acStdSig.bigWig 1\n",
      "cell: Same\n",
      "type: Same\n",
      "dataType: Same\n",
      "controlId: Different\n",
      "dccAccession: Same\n",
      "expId: Not shared.\n",
      "origAssembly: Same\n",
      "view: Same\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "lab: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "control: Same\n",
      "dateSubmitted: Same\n",
      "wgEncodeBroadHistoneNhekH3k4me3StdSig.bigWig 0\n",
      "control: Same\n",
      "controlId: Different\n",
      "project: Not shared.\n",
      "type: Same\n",
      "origAssembly: Same\n",
      "lab: Same\n",
      "dccAccession: Same\n",
      "cell: Same\n",
      "tableName: Not shared.\n",
      "treatment: Not shared.\n",
      "composite: Not shared.\n",
      "dataType: Same\n",
      "geoSampleAccession: Not shared.\n",
      "view: Same\n",
      "dateUnrestricted: Not shared.\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "md5sum: Not shared.\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "dateSubmitted: Same\n",
      "wgEncodeBroadHistoneNhekH3k4me3StdSig.bigWig 1\n",
      "cell: Same\n",
      "type: Same\n",
      "dataType: Same\n",
      "controlId: Different\n",
      "dccAccession: Same\n",
      "expId: Not shared.\n",
      "origAssembly: Same\n",
      "view: Same\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "lab: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "control: Same\n",
      "dateSubmitted: Same\n",
      "wgEncodeBroadHistoneK562H3k4me3StdSig.bigWig 0\n",
      "control: Same\n",
      "controlId: Different\n",
      "project: Not shared.\n",
      "type: Same\n",
      "origAssembly: Same\n",
      "lab: Same\n",
      "dccAccession: Same\n",
      "cell: Same\n",
      "tableName: Not shared.\n",
      "treatment: Not shared.\n",
      "composite: Not shared.\n",
      "dataType: Same\n",
      "geoSampleAccession: Not shared.\n",
      "view: Same\n",
      "dateUnrestricted: Not shared.\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "md5sum: Not shared.\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "dateSubmitted: Same\n",
      "wgEncodeBroadHistoneK562H3k4me3StdSig.bigWig 1\n",
      "cell: Same\n",
      "type: Same\n",
      "dataType: Same\n",
      "controlId: Different\n",
      "dccAccession: Same\n",
      "expId: Not shared.\n",
      "origAssembly: Same\n",
      "view: Same\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "lab: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "control: Same\n",
      "dateSubmitted: Same\n",
      "wgEncodeBroadHistoneGm12878H3k4me3StdSig.bigWig 0\n",
      "control: Same\n",
      "controlId: Different\n",
      "project: Not shared.\n",
      "type: Same\n",
      "objStatus: Not shared.\n",
      "origAssembly: Same\n",
      "lab: Same\n",
      "dccAccession: Same\n",
      "cell: Same\n",
      "tableName: Not shared.\n",
      "treatment: Not shared.\n",
      "composite: Not shared.\n",
      "dataType: Same\n",
      "md5sum: Not shared.\n",
      "view: Same\n",
      "dateUnrestricted: Not shared.\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "dateSubmitted: Same\n",
      "wgEncodeBroadHistoneGm12878H3k4me3StdSig.bigWig 1\n",
      "cell: Same\n",
      "type: Same\n",
      "dataType: Same\n",
      "controlId: Different\n",
      "dccAccession: Same\n",
      "expId: Not shared.\n",
      "origAssembly: Same\n",
      "view: Same\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "lab: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "control: Same\n",
      "dateSubmitted: Same\n",
      "wgEncodeBroadHistoneNhekH3k27acStdSig.bigWig 0\n",
      "control: Same\n",
      "controlId: Different\n",
      "project: Not shared.\n",
      "type: Same\n",
      "origAssembly: Same\n",
      "lab: Same\n",
      "dccAccession: Same\n",
      "cell: Same\n",
      "tableName: Not shared.\n",
      "treatment: Not shared.\n",
      "composite: Not shared.\n",
      "dataType: Same\n",
      "geoSampleAccession: Not shared.\n",
      "view: Same\n",
      "dateUnrestricted: Not shared.\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "md5sum: Not shared.\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "dateSubmitted: Same\n",
      "wgEncodeBroadHistoneNhekH3k27acStdSig.bigWig 1\n",
      "cell: Same\n",
      "type: Same\n",
      "dataType: Same\n",
      "controlId: Different\n",
      "dccAccession: Same\n",
      "expId: Not shared.\n",
      "origAssembly: Same\n",
      "view: Same\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "lab: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "control: Same\n",
      "dateSubmitted: Same\n",
      "wgEncodeBroadHistoneHsmmH3k27acStdSig.bigWig 0\n",
      "control: Same\n",
      "controlId: Different\n",
      "project: Not shared.\n",
      "type: Same\n",
      "origAssembly: Same\n",
      "lab: Same\n",
      "dccAccession: Same\n",
      "cell: Same\n",
      "tableName: Not shared.\n",
      "treatment: Not shared.\n",
      "composite: Not shared.\n",
      "dataType: Same\n",
      "geoSampleAccession: Not shared.\n",
      "view: Same\n",
      "dateUnrestricted: Not shared.\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "md5sum: Not shared.\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "dateSubmitted: Same\n",
      "wgEncodeBroadHistoneHsmmH3k27acStdSig.bigWig 1\n",
      "cell: Same\n",
      "type: Same\n",
      "dataType: Same\n",
      "controlId: Different\n",
      "dccAccession: Same\n",
      "expId: Not shared.\n",
      "origAssembly: Same\n",
      "view: Same\n",
      "dataVersion: Same\n",
      "setType: Same\n",
      "antibody: Same\n",
      "dateResubmitted: Same\n",
      "lab: Same\n",
      "subId: Same\n",
      "size: Same\n",
      "grant: Same\n",
      "control: Same\n",
      "dateSubmitted: Same\n"
     ]
    }
   ],
   "source": [
    "for file_name in hg19_meta_list:\n",
    "    if len(hg19_meta_list[file_name]) > 1:\n",
    "        for i in range(len(hg19_meta_list[file_name])):\n",
    "            print(file_name, end = \" \")\n",
    "            print(i)\n",
    "            for entry in hg19_meta_list[file_name][i]:\n",
    "                print(entry, end = \"\")\n",
    "                for j in range(len(hg19_meta_list[file_name])):\n",
    "                    if i == j:\n",
    "                        pass\n",
    "                    else:\n",
    "                        if entry in hg19_meta_list[file_name][j]:\n",
    "                            if hg19_meta_list[file_name][j][entry] == hg19_meta_list[file_name][i][entry]:\n",
    "                                print(\": Same\")\n",
    "                            else:\n",
    "                                print(\": Different\")\n",
    "                        else:\n",
    "                            print(\": Not shared.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "for file_name in hg19_meta_list:\n",
    "    if len(hg19_meta_list[file_name]) > 1:\n",
    "        new_list = []\n",
    "        for entry_dict in hg19_meta_list[file_name]:\n",
    "            if \"md5sum\" in entry_dict:\n",
    "                new_list.append(entry_dict)\n",
    "        hg19_meta_list[file_name] = new_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n"
     ]
    }
   ],
   "source": [
    "print(checkUniqueFileNames(hg19_meta))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read fields.  If duplicate, check for \"dateResubmitted\".  Remove entries with objStatus field."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def readMetaData(meta_file):\n",
    "    f = open(meta_file)\n",
    "    meta_dict = dict()\n",
    "    for line in f:\n",
    "        file_name = line.split(\"\\t\")[0]\n",
    "        #if file_name not in meta_dict:\n",
    "        #    meta_dict[file_name] = []\n",
    "        entry_dict = dict()\n",
    "        for entry in line.split(\"\\t\")[1].split(\";\"):\n",
    "            entry_dict[entry.strip().split(\"=\")[0]] = entry.strip().split(\"=\")[1]\n",
    "        #if \"md5sum\" in entry_dict:\n",
    "        #    meta_dict[file_name] = entry_dict\n",
    "        if file_name in meta_dict:\n",
    "            if \"dateResubmitted\" in entry_dict:\n",
    "                meta_dict[file_name] = entry_dict\n",
    "        else:\n",
    "            meta_dict[file_name] = entry_dict\n",
    "    return meta_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def getKeys(meta_dict):\n",
    "    key_list = []\n",
    "    for file_name in meta_dict:\n",
    "        for key in meta_dict[file_name]:\n",
    "            if key not in key_list:\n",
    "                key_list.append(key)\n",
    "    return key_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "hg19_meta = readMetaData(\"hg19_metadata.txt\")\n",
    "mm9_meta = readMetaData(\"mm9_metadata.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['labExpId', 'dccAccession', 'localization', 'tableName', 'composite', 'seqPlatform', 'bioRep', 'readType', 'protocol', 'replicate', 'lab', 'cell', 'rnaExtract', 'type', 'dataType', 'geoSampleAccession', 'view', 'dateUnrestricted', 'dataVersion', 'md5sum', 'subId', 'size', 'grant', 'project', 'dateSubmitted', 'origAssembly', 'treatment', 'labVersion', 'controlId', 'setType', 'antibody', 'control', 'mapAlgorithm', 'objStatus', 'submittedDataVersion', 'insertLength', 'dateResubmitted', 'spikeInPool', 'donorId', 'phase', 'sex', 'softwareVersion', 'expId', 'labProtocolId', 'platform', 'sourceObj', 'uniqueness']\n"
     ]
    }
   ],
   "source": [
    "print(getKeys(hg19_meta))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['strain', 'age', 'controlId', 'project', 'subId', 'tissueSourceType', 'softwareVersion', 'dataType', 'control', 'dccAccession', 'cell', 'tableName', 'type', 'composite', 'sex', 'labVersion', 'geoSampleAccession', 'setType', 'dateUnrestricted', 'dataVersion', 'md5sum', 'antibody', 'lab', 'view', 'size', 'grant', 'seqPlatform', 'dateSubmitted', 'treatment', 'replicate', 'labExpId', 'localization', 'readType', 'dateResubmitted', 'rnaExtract', 'protocol', 'bioRep', 'spikeInPool', 'platform', 'submittedDataVersion', 'objStatus', 'mapAlgorithm', 'insertLength', 'uniqueness', 'origAssembly']\n"
     ]
    }
   ],
   "source": [
    "print(getKeys(mm9_meta))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "revoked - This is a bigWig version 3 file, it is obsolete.\n",
      "renamed\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "renamed\n",
      "renamed\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "renamed\n",
      "replaced\n",
      "replaced\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "renamed\n",
      "revoked - data is mixed and merged with other experiments\n",
      "renamed\n",
      "replaced\n",
      "renamed\n",
      "renamed\n",
      "renamed\n",
      "renamed\n",
      "renamed\n",
      "renamed\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "replaced\n",
      "replaced by a Rep1V2 object\n",
      "replaced\n",
      "revoked - remapped entire cell line to correct sex, see V2\n",
      "revoked - Replaced by wgEncodeEH002836\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "revoked - Original experiment was mapped to wrong sex\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "revoked - remapped entire cell line to correct sex, see V2\n",
      "revoked - data is mixed and merged with other experiments\n",
      "replaced - A newer experiment has been run on the HiSeq2000\n",
      "renamed\n",
      "replaced\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "renamed\n",
      "revoked - data is mixed and merged with other experiments\n",
      "replaced\n",
      "replaced\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "replaced\n",
      "renamed - Files were renamed due to cell naming from K562b -> K562\n",
      "renamed\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "revoked - data is mixed and merged with other experiments\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "replaced - A newer experiment has been run on the HiSeq2000\n",
      "renamed - Files were renamed due to cell naming from K562b -> K562\n",
      "replaced\n",
      "renamed\n",
      "replaced\n",
      "renamed\n",
      "revoked - data is mixed and merged with other experiments\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "revoked - Original experiment was mapped to wrong sex\n",
      "replaced\n",
      "revoked - Lab decided to re-run this experiment\n",
      "replaced\n",
      "replaced - Replaced by wgEncodeEH002806, since this experiment which was relatively undersequenced due to technological limitations at the time it was completed.\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "replaced\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "replaced\n",
      "renamed\n",
      "renamed\n",
      "replaced\n",
      "revoked - data is mixed and merged with other experiments\n",
      "revoked - data is mixed and merged with other experiments\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "revoked - data is mixed and merged with other experiments\n",
      "replaced\n",
      "replaced - lab replaced this version with a V2\n",
      "replaced\n",
      "replaced\n",
      "renamed\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "replaced\n",
      "renamed\n",
      "revoked - data is mixed and merged with other experiments\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "replaced\n",
      "renamed\n",
      "replaced\n",
      "revoked - data is mixed and merged with other experiments\n",
      "revoked - old version of bigWig (version 3)\n",
      "revoked - unused control\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "renamed\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "renamed\n",
      "replaced\n",
      "revoked - data is mixed and merged with other experiments\n",
      "replaced - too shallow\n",
      "renamed\n",
      "renamed\n",
      "replaced by a Rep1V2 object\n",
      "renamed\n",
      "revoked - data is mixed and merged with other experiments\n",
      "renamed\n",
      "revoked - data is mixed and merged with other experiments\n",
      "replaced - The lab redid the experiment with better results, look for the V2 version of this file.\n",
      "renamed\n",
      "renamed\n",
      "replaced\n",
      "renamed\n",
      "replaced\n",
      "revoked - lab requested the data be pulled due to a problem with DNA digest leading to spurious signal spikes(noise)\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "revoked - The lab and working group have little confidence that the antibody used is actually HEY1, it looks like Pol2\n",
      "renamed\n",
      "renamed - antibody was missing the vendorCode data is the same\n",
      "renamed\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "replaced\n",
      "renamed\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "renamed\n",
      "revoked - Original experiment was mapped to wrong sex\n",
      "replaced\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "renamed\n",
      "revoked - data is mixed and merged with other experiments\n",
      "replaced\n",
      "revoked - POLR2A antibody failed to validate\n",
      "renamed\n",
      "renamed\n",
      "renamed\n",
      "revoked - data is mixed and merged with other experiments\n",
      "renamed\n",
      "revoked - Deemed to be too poor of quality by lab\n",
      "renamed - Cell renaming caused file to be renamed to wgEncodeOpenChromFaireNhaBaseOverlapSignal.bigwig\n",
      "revoked - The lab and working group have little confidence that the antibody used is actually HEY1, it looks like Pol2\n",
      "renamed\n",
      "revoked - data is mixed and merged with other experiments\n",
      "renamed\n",
      "revoked - Original experiment was mapped to wrong sex\n",
      "renamed\n",
      "revoked - The lab and working group have little confidence that the antibody used is actually HEY1, it looks like Pol2\n",
      "revoked - Original experiment was mapped to wrong sex\n",
      "renamed\n",
      "revoked - data is mixed and merged with other experiments\n",
      "revoked - treatment name is not included in the object name\n",
      "renamed\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "renamed\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "renamed\n",
      "renamed - added in cell donor name - wgEncodeOpenChromChipMonocd14ro01746InputSig\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "renamed\n",
      "revoked - data is mixed and merged with other experiments\n",
      "revoked - data is mixed and merged with other experiments\n",
      "renamed\n",
      "renamed - Files were renamed due to cell naming from K562b -> K562\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "revoked - data is mixed and merged with other experiments\n",
      "renamed\n",
      "replaced\n",
      "replaced\n",
      "replaced\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "renamed\n",
      "revoked - data is mixed and merged with other experiments\n",
      "renamed\n",
      "replaced - A newer experiment has been run on the HiSeq2000\n",
      "renamed - renamed to wgEncodeHaibTfbsK562E2f6V0416102RawRep2.bigWig\n",
      "renamed\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "renamed\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "replaced - A newer experiment has been run on the HiSeq2000\n",
      "replaced\n",
      "renamed\n",
      "renamed\n",
      "replaced\n",
      "renamed - renamed to wgEncodeHaibTfbsK562E2f6V0416102RawRep1.bigWig\n",
      "replaced\n",
      "revoked - poor quality\n",
      "replaced - A newer experiment has been run on the HiSeq2000\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "replaced - A newer experiment has been run on the HiSeq2000\n",
      "renamed - antibody was missing the vendorCode data is the same\n",
      "replaced\n",
      "revoked - Original experiment was mapped to wrong sex\n",
      "renamed\n",
      "revoked - Deemed to be too poor of quality by lab\n",
      "renamed\n",
      "renamed\n",
      "revoked - lab wanted to pull data due to a problem with DNA digest leading to spurious signal spikes(noise)\n",
      "renamed\n",
      "renamed\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "revoked - poor quality\n",
      "revoked - data is mixed and merged with other experiments\n",
      "revoked - data is mixed and merged with other experiments\n",
      "revoked - data is mixed and merged with other experiments\n",
      "renamed\n",
      "renamed\n",
      "replaced - A newer experiment has been run on the HiSeq2000\n",
      "replaced\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "renamed\n",
      "renamed - Files were renamed due to cell naming from K562b -> K562\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "revoked - lab requested the data be pulled due to a problem with DNA digest leading to spurious signal spikes(noise)\n",
      "replaced\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "renamed\n",
      "renamed - Files were renamed due to cell naming from K562b -> K562\n",
      "replaced\n",
      "renamed\n",
      "replaced\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "revoked - data is mixed and merged with other experiments\n",
      "renamed\n",
      "replaced\n",
      "replaced - Reanalyzed after adding in 3rd replicate\n",
      "revoked - POLR2A antibody failed to validate\n",
      "replaced - A newer experiment has been run on the HiSeq2000\n",
      "renamed\n",
      "renamed\n",
      "renamed\n",
      "renamed\n",
      "replaced\n",
      "renamed\n",
      "renamed\n",
      "renamed\n",
      "revoked - data is mixed and merged with other experiments\n",
      "renamed\n",
      "replaced - object is identifying wrong peaks\n",
      "renamed\n",
      "renamed\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "renamed\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "renamed\n",
      "renamed\n",
      "revoked - data is mixed and merged with other experiments\n",
      "replaced - The lab redid the experiment with better results, look for the V2 version of this file.\n",
      "renamed\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "revoked - DNA present in sample\n",
      "renamed\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "renamed\n",
      "renamed\n",
      "renamed - File got renamed since more clarity of the Cell line was determined.\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "revoked - Original experiment was mapped to wrong sex\n",
      "replaced\n",
      "renamed\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "revoked - treatment name is not included in the object name\n",
      "renamed\n",
      "renamed\n",
      "revoked - DNA present in sample\n",
      "revoked - data is mixed and merged with other experiments\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "replaced\n",
      "revoked - Original experiment was mapped to wrong sex\n",
      "renamed\n",
      "renamed\n",
      "revoked - data is mixed and merged with other experiments\n",
      "replaced\n",
      "revoked - POLR2A antibody failed to validate\n",
      "replaced\n",
      "renamed\n",
      "replaced\n",
      "renamed\n",
      "replaced\n",
      "renamed - Cell renaming caused file to be renamed to wgEncodeOpenChromFaireNhaSig.bigWig\n",
      "revoked - The lab and working group have little confidence that the antibody used is actually HEY1, it looks like Pol2\n",
      "replaced\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "revoked - POLR2A antibody failed to validate\n",
      "replaced\n",
      "renamed\n",
      "renamed\n",
      "renamed\n",
      "replaced\n",
      "revoked - Lab decided to re-run this experiment\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "replaced\n",
      "renamed\n",
      "revoked - Deemed to be too poor of quality by lab\n",
      "renamed\n",
      "replaced\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "renamed\n",
      "replaced - too shallow\n",
      "renamed\n",
      "renamed\n",
      "revoked - data is mixed and merged with other experiments\n",
      "revoked - Deemed to be too poor of quality by lab\n",
      "renamed\n",
      "renamed\n",
      "replaced\n",
      "renamed\n",
      "revoked - data is mixed and merged with other experiments\n",
      "replaced - Any previous versions of the data are bad, please use the V4 version.\n",
      "revoked - data is mixed and merged with other experiments\n",
      "renamed\n",
      "replaced - Replaced by wgEncodeEH002836\n",
      "renamed\n",
      "replaced\n"
     ]
    }
   ],
   "source": [
    "for file_name in hg19_meta:\n",
    "    if \"objStatus\" in hg19_meta[file_name]:\n",
    "        print(hg19_meta[file_name][\"objStatus\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "renamed - file was renamed to Wbrain\n",
      "renamed - file was renamed to Wbrain\n",
      "replaced - Replaced with better version\n",
      "replaced - New version of pooled data, which includes replicate 3 of data set\n",
      "renamed - Antibody renaming and changing treatment abbreviation caused file to be renamed to wgEncodeSydhTfbsMelCtcfsc15914Dm2p5dIggyaleSig.bigWig\n",
      "renamed - file was renamed to Wbrain\n",
      "renamed - file was renamed to Wbrain\n",
      "renamed - file was renamed to Wbrain\n",
      "renamed - file was renamed to Wbrain\n",
      "renamed - file was renamed to Wbrain\n",
      "renamed - Antibody renaming caused file to be renamed to wgEncodeSydhTfbsMelP300sc584IggrabSig.bigWig\n",
      "renamed - Antibody renaming caused file to be renamed to wgEncodeSydhTfbsMelCmybsc7874IggrabSig.bigWig\n",
      "renamed - file was renamed to Wbrain\n",
      "replaced - This experiment has been replaced with experiment wgEncodeEM002781\n",
      "replaced - 1 bigWig contained in tar ball was empty\n",
      "revoked - Only the input (control) datasets are submitted, no corresponding Tfbs datasets are submitted\n",
      "renamed - Changing treatment abbreviation caused file to be renamed to wgEncodeSydhTfbsMelInputDm2p5dIggyaleSig.bigWig\n",
      "revoked - Incorrect antibody used\n",
      "renamed - Changing treatment abbreviation caused file to be renamed to wgEncodeSydhTfbsMelRad21Dm2p5dIggrabSig.bigWig\n",
      "renamed - file was renamed to Wbrain\n"
     ]
    }
   ],
   "source": [
    "for file_name in mm9_meta:\n",
    "    if \"objStatus\" in mm9_meta[file_name]:\n",
    "        print(mm9_meta[file_name][\"objStatus\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def removeObjStatusEntries(meta_dict):\n",
    "    new_dict = dict()\n",
    "    for file_name in meta_dict:\n",
    "        if \"objStatus\" not in meta_dict[file_name]:\n",
    "            new_dict[file_name] = meta_dict[file_name]\n",
    "    return new_dict\n",
    "\n",
    "hg19_meta = removeObjStatusEntries(hg19_meta)\n",
    "mm9_meta = removeObjStatusEntries(mm9_meta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "False\n"
     ]
    }
   ],
   "source": [
    "print(\"objStatus\" in getKeys(hg19_meta))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "False\n"
     ]
    }
   ],
   "source": [
    "print(\"objStatus\" in getKeys(mm9_meta))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Associate stranded bigwigs together"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['tableName', 'view', 'md5sum', 'size']\n"
     ]
    }
   ],
   "source": [
    "different_keys = []\n",
    "for file_name in hg19_meta:\n",
    "    if \"Plus\" in file_name:\n",
    "        minus_name = file_name.replace(\"Plus\",\"Minus\")\n",
    "        for entry in hg19_meta[file_name]:\n",
    "            if entry in hg19_meta[minus_name]:\n",
    "                if hg19_meta[file_name][entry] != hg19_meta[minus_name][entry]:\n",
    "                    if entry not in different_keys:\n",
    "                        different_keys.append(entry)\n",
    "            else:\n",
    "                if entry not in different_keys:\n",
    "                        different_keys.append(entry)\n",
    "print(different_keys)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def associateStranded(meta_dict):\n",
    "    stranded_fields = ['tableName', 'view', 'md5sum', 'size']\n",
    "    associated_dict = dict()\n",
    "    for file_name in meta_dict:\n",
    "        if \"Plus\" in file_name:\n",
    "            entry_name = file_name.replace(\".bigWig\",\"\").replace(\"Plus\",\"\")\n",
    "            if entry_name not in associated_dict:\n",
    "                associated_dict[entry_name] = dict()\n",
    "            associated_dict[entry_name][\"plus_bigwig\"] = file_name\n",
    "            for field in stranded_fields:\n",
    "                if field in meta_dict[file_name]:\n",
    "                    associated_dict[entry_name]['plus_' + field] = meta_dict[file_name][field]\n",
    "            for field in meta_dict[file_name]:\n",
    "                if field not in stranded_fields:\n",
    "                    associated_dict[entry_name][field] = meta_dict[file_name][field]\n",
    "        elif \"Minus\" in file_name:\n",
    "            entry_name = file_name.replace(\".bigWig\",\"\").replace(\"Minus\",\"\")\n",
    "            if entry_name not in associated_dict:\n",
    "                associated_dict[entry_name] = dict()\n",
    "            associated_dict[entry_name][\"minus_bigwig\"] = file_name\n",
    "            for field in stranded_fields:\n",
    "                if field in meta_dict[file_name]:\n",
    "                    associated_dict[entry_name]['minus_' + field] = meta_dict[file_name][field]\n",
    "            for field in meta_dict[file_name]:\n",
    "                if field not in stranded_fields:\n",
    "                    associated_dict[entry_name][field] = meta_dict[file_name][field]\n",
    "        else:\n",
    "            entry_name = file_name.replace(\".bigWig\",\"\")\n",
    "            associated_dict[entry_name] = dict()\n",
    "            associated_dict[entry_name][\"ambig_bigwig\"] = file_name\n",
    "            for field in stranded_fields:\n",
    "                if field in meta_dict[file_name]:\n",
    "                    associated_dict[entry_name]['ambig_' + field] = meta_dict[file_name][field]\n",
    "            for field in meta_dict[file_name]:\n",
    "                if field not in stranded_fields:\n",
    "                    associated_dict[entry_name][field] = meta_dict[file_name][field]\n",
    "    return associated_dict\n",
    "\n",
    "hg19_final = associateStranded(hg19_meta) \n",
    "mm9_final = associateStranded(mm9_meta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4084\n",
      "1103\n",
      "3526\n",
      "976\n",
      "3526\n",
      "976\n"
     ]
    }
   ],
   "source": [
    "print(len(hg19_meta))\n",
    "print(len(mm9_meta))\n",
    "\n",
    "count = 0\n",
    "for file_name in hg19_meta:\n",
    "    if \"Minus\" not in file_name:\n",
    "        count += 1\n",
    "print(count)\n",
    "\n",
    "count = 0\n",
    "for file_name in mm9_meta:\n",
    "    if \"Minus\" not in file_name:\n",
    "        count += 1\n",
    "print(count)\n",
    "\n",
    "print(len(hg19_final))\n",
    "print(len(mm9_final))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for entry in mm9_final:\n",
    "    mm9_final[entry][\"genome_assembly\"] = \"mm9\"\n",
    "for entry in hg19_final:\n",
    "    hg19_final[entry][\"genome_assembly\"] = \"hg19\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "hg19_keys = getKeys(hg19_final)\n",
    "mm9_keys = getKeys(mm9_final)\n",
    "\n",
    "all_keys = hg19_keys + list(set(mm9_keys) - set(hg19_keys))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['labExpId', 'dccAccession', 'controlId', 'project', 'ambig_md5sum', 'ambig_view', 'treatment', 'genome_assembly', 'subId', 'protocol', 'replicate', 'lab', 'cell', 'type', 'ambig_bigwig', 'dataType', 'ambig_tableName', 'geoSampleAccession', 'setType', 'dateUnrestricted', 'dataVersion', 'ambig_size', 'antibody', 'composite', 'grant', 'dateSubmitted', 'origAssembly', 'labVersion', 'control', 'dateResubmitted', 'phase', 'plus_md5sum', 'localization', 'readType', 'plus_tableName', 'plus_bigwig', 'plus_view', 'minus_md5sum', 'rnaExtract', 'minus_tableName', 'minus_size', 'plus_size', 'donorId', 'bioRep', 'minus_view', 'minus_bigwig', 'seqPlatform', 'spikeInPool', 'sex', 'mapAlgorithm', 'platform', 'submittedDataVersion', 'insertLength', 'expId', 'labProtocolId', 'uniqueness', 'sourceObj', 'softwareVersion', 'age', 'strain', 'tissueSourceType']\n"
     ]
    }
   ],
   "source": [
    "print(all_keys)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "OUTPUT = open(\"combined_ENCODE_metadata.txt\", \"w\")\n",
    "\n",
    "for key in all_keys:\n",
    "    OUTPUT.write(\"\\t\" + key)\n",
    "OUTPUT.write(\"\\n\")\n",
    "\n",
    "def writeMetaData(meta_data, OUTPUT):\n",
    "    for entry in meta_data:\n",
    "        OUTPUT.write(entry)\n",
    "        for key in all_keys:\n",
    "            if key not in meta_data[entry]:\n",
    "                OUTPUT.write(\"\\t\")\n",
    "            else:\n",
    "                OUTPUT.write(\"\\t\" + meta_data[entry][key])\n",
    "        OUTPUT.write(\"\\n\")\n",
    "\n",
    "writeMetaData(hg19_final, OUTPUT)\n",
    "writeMetaData(mm9_final, OUTPUT)\n",
    "\n",
    "OUTPUT.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'cell': 'NHLF', 'type': 'bigWig', 'mapAlgorithm': 'TH1014', 'dataType': 'RnaSeq', 'localization': 'cell', 'view': 'Signal', 'rnaExtract': 'longPolyA', 'dateSubmitted': '2011-01-22', 'insertLength': '200', 'readType': '2x75', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'subId': '3413', 'sourceObj': 'wgEncodeRegTxnCaltechRnaSeqNhlfR2x75Il200SigRep1V2 wgEncodeRegTxnCaltechRnaSeqNhlfR2x75Il200SigRep2V2', 'size': '298M', 'grant': 'Myers', 'lab': 'Caltech', 'submittedDataVersion': 'V2 - A better mapping technique.'}\n",
      "{'cell': 'NHLF', 'type': 'bigWig', 'ambig_bigwig': 'wgEncodeRegTxnCaltechRnaSeqNhlfR2x75Il200SigPooled.bigWig', 'dataType': 'RnaSeq', 'localization': 'cell', 'rnaExtract': 'longPolyA', 'mapAlgorithm': 'TH1014', 'ambig_view': 'Signal', 'insertLength': '200', 'submittedDataVersion': 'V2 - A better mapping technique.', 'readType': '2x75', 'dataVersion': 'ENCODE Jan 2011 Freeze', 'ambig_size': '298M', 'genome_assembly': 'hg19', 'subId': '3413', 'sourceObj': 'wgEncodeRegTxnCaltechRnaSeqNhlfR2x75Il200SigRep1V2 wgEncodeRegTxnCaltechRnaSeqNhlfR2x75Il200SigRep2V2', 'grant': 'Myers', 'lab': 'Caltech', 'dateSubmitted': '2011-01-22'}\n"
     ]
    }
   ],
   "source": [
    "print(hg19_meta[\"wgEncodeRegTxnCaltechRnaSeqNhlfR2x75Il200SigPooled.bigWig\"])\n",
    "print(hg19_final[\"wgEncodeRegTxnCaltechRnaSeqNhlfR2x75Il200SigPooled\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
