# -*- coding: utf-8 -*-
# Generated by Django 1.9.6 on 2016-06-10 19:51
from __future__ import unicode_literals

import django.core.validators
from django.db import migrations, models
import django.db.models.deletion
import utils.models


class Migration(migrations.Migration):

    dependencies = [
        ('analysis', '0029_auto_20160519_1054'),
    ]

    operations = [
        migrations.AlterField(
            model_name='analysis',
            name='anchor',
            field=models.PositiveSmallIntegerField(choices=[(0, 'start'), (1, 'center'), (2, 'end')], default=1, help_text='Where to center analysis window relative to BED range'),
        ),
        migrations.AlterField(
            model_name='analysis',
            name='bin_number',
            field=models.PositiveIntegerField(default=50, help_text='Number of bins to use in search window', validators=[django.core.validators.MinValueValidator(50), django.core.validators.MaxValueValidator(250)]),
        ),
        migrations.AlterField(
            model_name='analysis',
            name='bin_size',
            field=models.PositiveIntegerField(default=100, help_text='Size of bins to use in search window', validators=[django.core.validators.MinValueValidator(1)]),
        ),
        migrations.AlterField(
            model_name='analysis',
            name='bin_start',
            field=models.IntegerField(default=-2500, help_text='Distance from anchor to start designating bins'),
        ),
        migrations.AlterField(
            model_name='analysisdatasets',
            name='count_matrix',
            field=models.ForeignKey(help_text='Matrix of read coverage over genomic features', null=True, on_delete=django.db.models.deletion.CASCADE, to='analysis.FeatureListCountMatrix'),
        ),
        migrations.AlterField(
            model_name='encodedataset',
            name='antibody',
            field=models.CharField(blank=True, db_index=True, help_text='Antibody used in pulldown; reported by ENCODE', max_length=32),
        ),
        migrations.AlterField(
            model_name='encodedataset',
            name='cell_type',
            field=models.CharField(db_index=True, help_text='Cell type (cell line or tissue); reported by ENCODE', max_length=32),
        ),
        migrations.AlterField(
            model_name='encodedataset',
            name='data_ambiguous',
            field=models.FileField(blank=True, help_text='Coverage data for which strand is ambiguous or unknown', max_length=256, storage=utils.models.ReadOnlyFileSystemStorage(location='/Users/shapiroaj4/dev/genomics/data/encode'), upload_to=''),
        ),
        migrations.AlterField(
            model_name='encodedataset',
            name='data_minus',
            field=models.FileField(blank=True, help_text='Coverage data for which strand is minus', max_length=256, storage=utils.models.ReadOnlyFileSystemStorage(location='/Users/shapiroaj4/dev/genomics/data/encode'), upload_to=''),
        ),
        migrations.AlterField(
            model_name='encodedataset',
            name='data_plus',
            field=models.FileField(blank=True, help_text='Coverage data for which strand is plus', max_length=256, storage=utils.models.ReadOnlyFileSystemStorage(location='/Users/shapiroaj4/dev/genomics/data/encode'), upload_to=''),
        ),
        migrations.AlterField(
            model_name='encodedataset',
            name='data_type',
            field=models.CharField(db_index=True, help_text='Experiment type', max_length=16),
        ),
        migrations.AlterField(
            model_name='encodedataset',
            name='localization',
            field=models.CharField(blank=True, db_index=True, help_text='Cellular localization; reported by ENCODE', max_length=32),
        ),
        migrations.AlterField(
            model_name='encodedataset',
            name='phase',
            field=models.CharField(blank=True, db_index=True, help_text='Cell phase; reported by ENCODE', max_length=32),
        ),
        migrations.AlterField(
            model_name='encodedataset',
            name='rna_extract',
            field=models.CharField(blank=True, db_index=True, help_text='RNA extraction protocol; reported by ENCODE', max_length=32),
        ),
        migrations.AlterField(
            model_name='encodedataset',
            name='treatment',
            field=models.CharField(blank=True, db_index=True, help_text='Experimental treatment of cells; reported by ENCODE', max_length=32),
        ),
        migrations.AlterField(
            model_name='featurelistcountmatrix',
            name='anchor',
            field=models.PositiveSmallIntegerField(choices=[(0, 'start'), (1, 'center'), (2, 'end')], default=1, help_text='Where to center analysis window relative to BED range'),
        ),
        migrations.AlterField(
            model_name='featurelistcountmatrix',
            name='bin_number',
            field=models.PositiveIntegerField(default=50, help_text='Number of bins to use in search window', validators=[django.core.validators.MinValueValidator(50), django.core.validators.MaxValueValidator(250)]),
        ),
        migrations.AlterField(
            model_name='featurelistcountmatrix',
            name='bin_size',
            field=models.PositiveIntegerField(default=100, help_text='Size of bins to use in search window', validators=[django.core.validators.MinValueValidator(1)]),
        ),
        migrations.AlterField(
            model_name='featurelistcountmatrix',
            name='bin_start',
            field=models.IntegerField(default=-2500, help_text='Distance from anchor to start designating bins'),
        ),
        migrations.AlterField(
            model_name='featurelistcountmatrix',
            name='matrix',
            field=models.FileField(help_text='Matrix file of read coverage', max_length=256, upload_to='fcm/'),
        ),
        migrations.AlterField(
            model_name='userdataset',
            name='ambiguous',
            field=models.ForeignKey(help_text='Coverage data for which strand is ambiguous or unknown', null=True, on_delete=django.db.models.deletion.CASCADE, related_name='ambiguous', to='analysis.DatasetDownload'),
        ),
        migrations.AlterField(
            model_name='userdataset',
            name='data_type',
            field=models.CharField(choices=[('Cage', 'Cage'), ('ChiaPet', 'ChiaPet'), ('ChipSeq', 'ChipSeq'), ('DnaseDgf', 'DnaseDgf'), ('DnaseSeq', 'DnaseSeq'), ('FaireSeq', 'FaireSeq'), ('Mapability', 'Mapability'), ('Nucleosome', 'Nucleosome'), ('Orchid', 'Orchid'), ('RepliChip', 'RepliChip'), ('RepliSeq', 'RepliSeq'), ('RipSeq', 'RipSeq'), ('RnaPet', 'RnaPet'), ('RnaSeq', 'RnaSeq'), ('SmartSeq', 'SmartSeq'), ('Other', 'Other (describe in "description" field)')], help_text='Experiment type', max_length=16),
        ),
        migrations.AlterField(
            model_name='userdataset',
            name='minus',
            field=models.ForeignKey(help_text='Coverage data for which strand is minus', null=True, on_delete=django.db.models.deletion.CASCADE, related_name='minus', to='analysis.DatasetDownload'),
        ),
        migrations.AlterField(
            model_name='userdataset',
            name='plus',
            field=models.ForeignKey(help_text='Coverage data for which strand is plus', null=True, on_delete=django.db.models.deletion.CASCADE, related_name='plus', to='analysis.DatasetDownload'),
        ),
    ]
